File: app/src/main/java/com/example/myappcontacts/presentation/contacts/view/ContactsFragment.java
package com.example.myappcontacts.presentation.contacts.view;

import android.content.Intent;
import android.graphics.Color;
import android.net.Uri;
import android.os.Bundle;
import android.provider.MediaStore;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.telephony.PhoneNumberFormattingTextWatcher;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.arellomobile.mvp.MvpAppCompatFragment;
import com.arellomobile.mvp.presenter.InjectPresenter;
import com.example.myappcontacts.R;
import com.example.myappcontacts.data.dao.contacts.db.ContactsModel;
import com.example.myappcontacts.presentation.contacts.presenter.ContactsPresenter;
import com.facebook.drawee.view.SimpleDraweeView;

import java.util.UUID;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;


public class ContactsFragment extends MvpAppCompatFragment implements IContactsView {
    public static final String ARG_CONTACT_ID = "contactId";

    private UUID mContactId;
    private Unbinder mUnbinder;

    @InjectPresenter
    ContactsPresenter mContactsPresenter;

    @BindView(R.id.contact_photo)
    SimpleDraweeView mPhotoImage;

    @BindView(R.id.contacts_text_views)
    LinearLayout mTextViews;
    @BindView(R.id.name_text_view)
    TextView mNameTextView;
    @BindView(R.id.company_text_view)
    TextView mCompanyTextView;
    @BindView(R.id.tel_number_text_view)
    TextView mTelNumberTextView;
    @BindView(R.id.address_text_view)
    TextView mAddressTextView;

    @BindView(R.id.contacts_edit_texts)
    LinearLayout mEditTexts;
    @BindView(R.id.first_name_edit_text)
    EditText mFirstNameEditText;
    @BindView(R.id.last_name_edit_text)
    EditText mLastNameEditText;
    @BindView(R.id.company_name_edit_text)
    EditText mCompanyNameEditText;
    @BindView(R.id.tel_number_edit_text)
    EditText mTelNumberEditText;
    @BindView(R.id.address_edit_text)
    TextView mAddressEditText;

    MenuItem mEdit_Save_Button;


    public ContactsFragment() {
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            mContactId = UUID.fromString(getArguments().getString(ARG_CONTACT_ID));
        }

    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_contact, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        setHasOptionsMenu(true);
        mUnbinder = ButterKnife.bind(this, view);
        mTelNumberEditText.addTextChangedListener(new PhoneNumberFormattingTextWatcher());
        mPhotoImage.setOnClickListener(v -> mContactsPresenter.onPhotoImageClicked());
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        inflater.inflate(R.menu.fragment_contacts_menu, menu);
        mEdit_Save_Button = menu.findItem(R.id.edit_save_changes);
        mContactsPresenter.loadContact(mContactId);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.edit_save_changes:
                mContactsPresenter.onEditSaveButtonClicked();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }

    }

    @Override
    public void updateUI(ContactsModel contactsModel, boolean edit) {
        Log.i("MY_TAG2", "updateUI сработал");

        updateMenuUI(edit);
        visibilityEdit(edit);
        if (edit) {
            editContact(contactsModel);
        } else {
            showContact(contactsModel);
        }
        if (contactsModel.getPhotoUri() != null) {
            mPhotoImage.setImageURI(Uri.parse(contactsModel.getPhotoUri()));
        }
    }

    private void visibilityEdit(boolean edit) {
        View[] editViews = {mEditTexts, mTelNumberEditText, mAddressEditText};
        View[] textViews = {mTextViews, mTelNumberTextView, mAddressTextView};
        int editViewVisibility = edit ? View.VISIBLE : View.GONE;
        int textViewVisibility = edit ? View.GONE : View.VISIBLE;
        for (View editView : editViews) {
            editView.setVisibility(editViewVisibility);
        }
        for (View textView :
                textViews) {
            textView.setVisibility(textViewVisibility);
        }
    }

    private void showContact(ContactsModel contactsModel) {
        String firstName = contactsModel.getFirstName();
        String lastName = contactsModel.getLastName();
        if (!firstName.equals("") || !lastName.equals("")) {
            mNameTextView.setText(getString(R.string.contact_name, contactsModel.getLastName(), contactsModel.getFirstName()));
        } else {
            mNameTextView.setText(getString(R.string.no_name));
        }
        String companyName = contactsModel.getCompanyName();
        if (companyName.equals("")) {
            mCompanyTextView.setVisibility(View.GONE);
        } else {
            mCompanyTextView.setVisibility(View.VISIBLE);
            mCompanyTextView.setText(contactsModel.getCompanyName());
        }
        String telNumber = contactsModel.getTelNumber();
        if (!telNumber.equals("")) {
            mTelNumberTextView.setText(contactsModel.getTelNumber());
            mTelNumberTextView.setTextColor(Color.BLACK);
        } else {
            mTelNumberTextView.setText(getString(R.string.no_telephone));
            mTelNumberTextView.setTextColor(Color.RED);
        }
        String address = contactsModel.getAddress();
        if (address.equals("")) {
            mAddressTextView.setText(getString(R.string.no_address));
        } else {
            mAddressTextView.setText(address);
        }
    }

    private void editContact(ContactsModel contactsModel) {
        mFirstNameEditText.setText(contactsModel.getFirstName());
        mLastNameEditText.setText(contactsModel.getLastName());
        mCompanyNameEditText.setText(contactsModel.getCompanyName());
        mTelNumberEditText.setText(contactsModel.getTelNumber());
        mAddressEditText.setText(contactsModel.getAddress());
    }

    private void updateMenuUI(boolean edit) {
        if (mEdit_Save_Button != null) {
            if (!edit) {
                mEdit_Save_Button.setTitle(R.string.edit_contact);
                mEdit_Save_Button.setIcon(getResources().getDrawable(R.drawable.ic_edit_contact));
            } else {
                mEdit_Save_Button.setIcon(getResources().getDrawable(R.drawable.ic_save_changes));
                mEdit_Save_Button.setTitle(R.string.save_contact);
            }
        }
    }

    @Override
    public void updateContact() {
        if (mTelNumberEditText.getText().toString().equals("")) {
            mTelNumberEditText.setHintTextColor(Color.RED);
        } else {
            ContactsModel contactsModel = new ContactsModel(mContactId);
            contactsModel.setFirstName(mFirstNameEditText.getText().toString());
            contactsModel.setLastName(mLastNameEditText.getText().toString());
            contactsModel.setCompanyName(mCompanyNameEditText.getText().toString());
            contactsModel.setTelNumber(mTelNumberEditText.getText().toString());
            contactsModel.setAddress(mAddressEditText.getText().toString());
            mContactsPresenter.updateContact(contactsModel);
        }
    }

    @Override
    public void updatePhoto() {
        startActivityForResult(new Intent(Intent.ACTION_PICK, MediaStore.Images.Media.EXTERNAL_CONTENT_URI), 100);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == 100 && data != null) {
            Uri uri = data.getData();
            mPhotoImage.setImageURI(uri);
            mContactsPresenter.photoUriLoaded(uri.toString());
        }
    }

    @Override
    public void onDestroy() {
        mUnbinder.unbind();
        super.onDestroy();
    }
}

File: app/src/main/java/com/example/myappcontacts/presentation/contacts/view/ContactsFragment.kt
package com.example.myappcontacts.presentation.contacts.view

import android.content.Intent
import android.graphics.Color
import android.os.Bundle
import android.provider.MediaStore
import android.support.v4.content.ContextCompat
import android.telephony.PhoneNumberFormattingTextWatcher
import android.view.*
import com.arellomobile.mvp.MvpAppCompatFragment
import com.arellomobile.mvp.presenter.InjectPresenter
import com.example.myappcontacts.R
import com.example.myappcontacts.data.dao.contacts.db.ContactsModel
import com.example.myappcontacts.presentation.contacts.presenter.ContactsPresenter
import kotlinx.android.synthetic.main.fragment_contact.*
import java.util.*

const val ARG_CONTACT_ID = "contactId"
const val REQUEST_CODE_PHOTO = 100

class ContactsFragment : MvpAppCompatFragment(), IContactsView {

    @InjectPresenter
    lateinit var contactsPresenter: ContactsPresenter
    lateinit var contactId: UUID

    private var editSaveButton: MenuItem? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (arguments!!.containsKey(ARG_CONTACT_ID)) {
            contactId = UUID.fromString(arguments?.getString(ARG_CONTACT_ID))
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        return inflater.inflate(R.layout.fragment_contact, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setHasOptionsMenu(true)
        tel_number_edit_text.addTextChangedListener(PhoneNumberFormattingTextWatcher())
        contact_photo.setOnClickListener { contactsPresenter.onPhotoImageClicked() }
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater.inflate(R.menu.fragment_contacts_menu, menu)

        editSaveButton = menu.findItem(R.id.edit_save_changes)

        contactsPresenter.loadContact(contactId)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean =
            when (item.itemId) {
                R.id.edit_save_changes -> {
                    contactsPresenter.onEditSaveButtonClicked()
                    true
                }
                else -> super.onOptionsItemSelected(item)
            }

    override fun updateUI(contactsModel: ContactsModel, edit: Boolean) {
        updateMenuUI(edit)
        visibilityEdit(edit)
        if (edit) {
            editContact(contactsModel)
        } else {
            showContact(contactsModel)
        }
        if (contactsModel.photoUri.isNotBlank()) {
            contact_photo.setImageURI(contactsModel.photoUri)
        }
    }

    private fun visibilityEdit(edit: Boolean) {
        val editViews = arrayOf(contacts_edit_texts, tel_number_edit_text, address_edit_text)
        val textViews = arrayOf(contacts_text_views, tel_number_text_view, address_text_view)
        val editViewVisibility = if (edit) View.VISIBLE else View.GONE
        val textViewVisibility = if (edit) View.GONE else View.VISIBLE
        editViews.forEach { it.visibility = editViewVisibility }
        textViews.forEach { it.visibility = textViewVisibility }
    }

    private fun showContact(contactsModel: ContactsModel) {
        with(contactsModel) {
            name_text_view.text =
                    if (firstName.isNotBlank() || lastName.isNotBlank()) {
                        getString(R.string.contact_name, firstName, lastName).trim()
                    } else {
                        getString(R.string.no_name)
                    }

            if (companyName.isNotBlank()) {
                company_text_view.visibility = View.VISIBLE
                company_text_view.text = contactsModel.companyName
            } else {
                company_text_view.visibility = View.GONE
            }

            if (telNumber.isNotBlank()) {
                tel_number_text_view.text = contactsModel.telNumber
                tel_number_text_view.setTextColor(Color.BLACK)
            } else {
                tel_number_text_view.text = getString(R.string.no_telephone)
                tel_number_text_view.setTextColor(Color.RED)
            }

            address_text_view.text =
                    if (address.isNotBlank()) address
                    else getString(R.string.no_address)

        }
    }

    private fun editContact(contactsModel: ContactsModel) {
        with(contactsModel) {
            first_name_edit_text.setText(firstName)
            last_name_edit_text.setText(lastName)
            company_name_edit_text.setText(companyName)
            tel_number_edit_text.setText(telNumber)
            address_edit_text.setText(address)
        }
    }

    private fun updateMenuUI(edit: Boolean) {
        if (edit) {
            editSaveButton?.setTitle(R.string.save_contact)
            editSaveButton?.icon = ContextCompat.getDrawable(context!!, R.drawable.ic_save_changes)
        } else {
            editSaveButton?.setTitle(R.string.edit_contact)
            editSaveButton?.icon = ContextCompat.getDrawable(context!!, R.drawable.ic_edit_contact)
        }
    }

    override fun updateContact() {
        if (tel_number_edit_text.text.toString().isBlank()) {
            tel_number_edit_text.setHintTextColor(Color.RED)
        } else {
                contactsPresenter.updateContact(saveContactsModel())
            }
        }

    private fun saveContactsModel(): ContactsModel {
        with(ContactsModel(contactId)) {
            firstName = first_name_edit_text.text.toString()
            lastName = last_name_edit_text.text.toString()
            companyName = company_name_edit_text.text.toString()
            telNumber = tel_number_edit_text.text.toString()
            address = address_edit_text.text.toString()

            contactsPresenter.saveContactsModel(this)
            return this
        }
    }

    override fun updatePhoto() {
        startActivityForResult(Intent(Intent.ACTION_PICK,
                MediaStore.Images.Media.EXTERNAL_CONTENT_URI),
                REQUEST_CODE_PHOTO)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == REQUEST_CODE_PHOTO) {
            val uri = data?.data ?: return
            contact_photo.setImageURI(uri.toString())
            contactsPresenter.photoUriLoaded(uri.toString())
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        saveContactsModel()
    }
}
